docs/troubleshooting.md

```diff
@@ -0,0 +1,99 @@
+
+This is the case when you write a `config.js` file and pass it to a tool.
+
+> This is the "ideal" way for JavaScript libraries to be configured, because it allows an easy
+> "symmetry" where you `require` a library and declare your dependency on it in the same place.
+
+In this case you should add the runtime dependency to your BUILD file where the `config.js` is a source.
+
+For example,
+
+```starlark
+js_library(
+    name = "requires_foo",
+    srcs = ["config.js"],          # contains "require('foo')"
+    data = [":node_modules/foo"],  # satisfies that require
+)
+```
+
+### require appears in third-party code
+
+This case itself breaks down into three possible remedies, depending on whether you can move the
+require to your own code, the missing dependency can be considered a "bug",
+or the third-party package uses the "plugin pattern" to discover its
+plugins dynamically at runtime based on finding them based on a string you provided.
+
+#### The `require` can move to first-party
+
+This is the most principled solution. In many cases, a library that accepts the name of a package as
+a string will also accept it as an object, so you can refactor `config: ['some-package']` to
+`config: [require('some-package')]`. You may need to change from json or yaml config to a JavaScript
+config file to allow the `require` syntax.
+
+Once you've done this, it's handled like the "require appears in your code" case above.
+
+For example, the
+[documentation for the postcss-loader for Webpack](https://webpack.js.org/loaders/postcss-loader/#sugarss)
+suggests that you `npm install --save-dev sugarss`
+and then pass the string "sugarss" to the `options.postcssOptions.parser` property of the loader.
+However this violates symmetry and would require workarounds listed below.
+You can simply pass `require("sugarss")` instead of the bare string, then include the `sugarss`
+package in the `data` (runtime dependencies) of your `webpack.config.js`.
+
+#### It's a bug
+
+This is the case when a package has a `require` statement in its runtime code for some package, but
+it doesn't list that package in its `package.json`, or lists it only as a `devDependency`.
+
+pnpm and Yarn PnP will hit the same bug. Conveniently, there's already a shared database used by
+both projects to list these, along with the missing dependency edge:
+https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-extensions/sources/index.ts
+
+> We should use this database under Bazel as well. Follow
+> https://github.com/aspect-build/rules_js/issues/1215.
+
+The recommended fix for both pnpm and rules_js is to use
+[pnpm.packageExtensions](https://pnpm.io/package_json#pnpmpackageextensions)
+in your `package.json` to add the missing `dependencies` or `peerDependencies`.
+
+Example,
+
+https://github.com/aspect-build/rules_js/blob/a8c192eed0e553acb7000beee00c60d60a32ed82/package.json#L12
+
+Make sure you run `pnpm install` after changing `package.json`, as rules_js only reads the
```

Comment:
Mention `--lockfile-only` here since it is faster? Also mention how to run the bazel provisioned pnpm?

Reply:
pointed to https://docs.aspect.build/rules/aspect_rules_js/docs/#fetch-third-party-packages-from-npm which now has the `--lockfile-only` flag, to avoid this repeating in many places